File: 2161-partition-array-according-to-given-pivot/2161-partition-array-according-to-given-pivot.py
class Solution:
    def pivotArray(self, nums: List[int], pivot: int) -> List[int]:
        l = [] # elements less than pivot
        e = [] # elements equal to pivot
        g = [] # elements greater than pivot
        for num in nums:
            if num < pivot:
                l.append(num)
            elif num > pivot:
                g.append(num)
            else:
                e.append(num)
        
        return l + e + g

# array
# time: O(n)
# space: O(n)

File: 2161-partition-array-according-to-given-pivot/2161-partition-array-according-to-given-pivot.go
func pivotArray(nums []int, pivot int) []int {
    less := []int{} // elements less than pivot
    equal := []int{} // element equal to pivot
    greater := []int{} // elements greater than pivot
    for _, num := range nums {
        if num < pivot {
            less = append(less, num)
        } else if num > pivot {
            greater = append(greater, num)
        } else {
            equal = append(equal, num)
        }
    }

    ans := make([]int, 0, len(nums)) // pre-allocate for efficiency
    ans = append(ans, less...)
    ans = append(ans, equal...)
    ans = append(ans, greater...)

    return ans
}

// array
// time: O(n)
// space: O(n)